data-prepper-core/src/integrationTest/java/org/opensearch/dataprepper/plugins/InMemorySourceAccessor.java

```diff
@@ -6,23 +6,84 @@
 package org.opensearch.dataprepper.plugins;
 
 import org.opensearch.dataprepper.model.event.Event;
+import org.opensearch.dataprepper.model.event.JacksonEvent;
 import org.opensearch.dataprepper.model.record.Record;
+import org.opensearch.dataprepper.model.event.EventFactory;
+import org.opensearch.dataprepper.model.event.EventBuilder;
 
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.HashMap;
 import java.util.List;
+import java.util.UUID;
 import java.util.Map;
 import java.util.concurrent.locks.Lock;
 import java.util.concurrent.locks.ReentrantLock;
+import java.util.concurrent.atomic.AtomicBoolean;
 
 /**
  * Provides a mechanism to write records to an in_memory source. This allows the pipeline to execute
  * independently of the test code, but the test code can write to a source.
  */
 public class InMemorySourceAccessor {
     private final Map<String, List<Record<Event>>> recordsMap = new HashMap<>();
-    private final Lock lock = new ReentrantLock();
+    private EventFactory eventFactory;
+    private AtomicBoolean ackReceived;
+    private final Lock lock;
+
+    public InMemorySourceAccessor() {
+        this.lock = new ReentrantLock();
+    }
+
+    public void setEventFactory(final EventFactory factory) {
+        eventFactory = factory;
+    }
+
+    public void setAckReceived(boolean newValue) {
+        if (ackReceived == null) {
+            ackReceived = new AtomicBoolean(newValue);
+        } else {
+            ackReceived.set(newValue);
+        }
+    }
+
+    public Boolean getAckReceived() {
+        if (ackReceived == null) {
+            return null;
+        }
+        return ackReceived.get();
+    }
+
+    public void submit(final String testingKey, int numRecords) {
+        if (eventFactory == null) {
+            return;
+        }
+        List<Record<Event>> records = new ArrayList<>();
+        for (int i = 0; i < numRecords; i++) {
+            Map<String, Object> eventMap = Map.of("message", UUID.randomUUID().toString());
+            EventBuilder eventBuilder = (EventBuilder) eventFactory.eventBuilder(EventBuilder.class).withData(eventMap);
+            JacksonEvent event = (JacksonEvent) eventBuilder.build();
+            records.add(new Record<Event>(event));
+        }
+        submit(testingKey, records);
+    }
+
+    public void submitWithStatus(final String testingKey, int numRecords) {
```

Comment:
I think you could submit the `List<Record<Event>>` to give the actual tests more control over the data and then also have an ability to verify the outcomes better.

Reply:
Yes. But I felt basic checking of the event is enough in this case because most important thing is that we received the expected number and the ack. I could add check for input records data and compare them with output records as a separate PR.